src/toxcore/dht/packet/cookie.rs

```diff
@@ -0,0 +1,227 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2018 Evgeny Kurnevsky <kurnevsky@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*! Cookie struct
+*/
+
+use nom::be_u64;
+
+use std::io::{Error, ErrorKind};
+
+use toxcore::binary_io::*;
+use toxcore::crypto_core::*;
+
+/**
+
+Serialized form:
+
+Length | Content
+------ | ------
+`8`    | Cookie timestamp
+`32`   | Long term `PublicKey`
+`32`   | DHT `PublicKey`
+
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Cookie {
+    /// Time when this cookie was generated
+    pub time: u64,
+    /// Long term `PublicKey`
+    pub real_pk: PublicKey,
+    /// DHT `PublicKey`
+    pub dht_pk: PublicKey,
+}
+
+impl FromBytes for Cookie {
+    named!(from_bytes<Cookie>, do_parse!(
+        time: be_u64 >>
+        real_pk: call!(PublicKey::from_bytes) >>
+        dht_pk: call!(PublicKey::from_bytes) >>
+        eof!() >>
```

Comment:
Why eof?

Reply:
c-toxcore checks exact length - if it's bigger then packet will be droppet. So I think we can be restrictive here too - it's better to drop suspicious packet then process them.

Reply:
ok